README.md
klingon-assistant
=================

**tlhIngan Hol boQwI'** (**boQwI'** for short) is a lexicon dictionary for the Klingon language (**tlhIngan Hol**) and a tool for analysing the grammar of Klingon words and sentences.

Please support the Klingon language by purchasing "The Klingon Dictionary", "Klingon for the Galactic Traveler", "The Klingon Way", "Conversational Klingon", "Power Klingon", and other Klingon- and Star Trek-related products from Pocket Books, Simon & Schuster, and Paramount/Viacom/CBS Entertainment.

Klingon, [Star Trek](http://startrek.com/), and related marks are trademarks of CBS Studios, Inc., and are used under "fair use" guidelines.

Get this app [on Google Play](https://play.google.com/store/apps/details?id=org.tlhInganHol.android.klingonassistant). An accompanying [Text-To-Speech Engine](https://play.google.com/store/apps/details?id=org.tlhInganHol.android.klingonttsengine) is also available.


KlingonAssistant/CHANGELOG
CHANGELOG

2018-02-25: Vocab from comments by Maltz about {boQwI'}.
2018-02-08: DSC vocab up to S01E15.
2018-01-29: some German translations of DSC vocab.
2018-01-11: Misc. edits, paq'batlh page numbers
2017-12-18: Misc. edits, some German translations.
2017-11-27: Misc. edits.
2017-11-13: Fix bug with link target on last line. Added qepHom'a' 2017 vocab.
2017-11-05: If KWOTD fails to match entry in db, treat as search. Lock bottom nav buttons to bottom of screen.
2017-10-31: More unconfirmed DSC words. Remove FAB, adjust UI spacing.
2017-10-29: Exclude extra section from random button.
2017-10-25: Add DSC font. Fix locale issues. Improve Preferences activity.
2017-10-24: Turn off copy and paste in entry. Hide KWOTD menu, and automatically start KWOTD job.
2017-10-20: Determined that ForegroundColorSpan fix is not cause of crash on Pixel, etc.
2017-10-17: Added KWOTD functionality. Changes to font family for prefix charts. Fix colour span bug (again).
2017-10-12: Added even more lesson functionality.
2017-10-10: Fix for multi-user database issue (#385) on devices with API 19+.
2017-10-09: Added more lesson functionality.
2017-10-05: Added lesson activity (hidden under preferences).
2017-10-01: Handle swipe to move between search results; put verb before verb stem in results
2017-09-25: Add opening monologue of DSC, make misc. edits related to DSC
2017-09-24: Fix tablet layout bug, make database IDs sequential without gaps, misc lint fixes
2017-09-20: allow edit of mistyped input, selection of text, search in {pIqaD}
2017-09-18: move v2 to main version; bug fixes
2017-09-16: alpha release of v2
2017-09-14: Target new version to API 26 and above. (For one, it allows us to not deal with the Jelly Bean span bug, which still exists. Also, "tlh" as a locale does not exist on older devices.)
2017-09-10: Migrated from ant to gradle (deprecate support of API below 16). Bug fixes. Enabled TTS always.
2017-09-10: Removed "float" mode. This was not being used and was confusing users.
2017-09-08: Misc. edits.
2017-08-20: Add (almost all) Klingon lines from movies as entries.
2017-08-14: Renumber entry IDs as part of clean-up to prep for DSC (Star Trek: Discovery) entries.
2017-08-09: Last version (maybe) before freezing development.
2017-08-06: Links and edits to weather words.
2017-08-03: More German notes translated. Fixed errors in a few entries.
2017-08-01: Add German examples. Add names from STLV. Fixed some typos.
2017-07-31: New words from qep'a' 2017 and Star Trek: Discovery, including German translations.
2017-07-25: Make German the primary language when option is selected.
2017-07-18: Enabled German search tags and notes in database.
2017-07-13: Added vocabulary about marriage.
2017-07-12: Added 2 entries from the {paq'batlh}.
2017-07-05: Edits to some {may'}-related entries. Edit notes on rovers on adjectives.
2017-05-29: Various minor edits.
2017-04-04: Finished entering German definitions.
2017-02-21: Minor edits.
2017-02-01: Added more definitions in German. Minor edits.
2016-12-01: Add vocabulary from Pop Culture Hero Coalition announcement.
2016-11-03: Add Halloween and qepHom vocabulary.
2016-10-20: Add info on {jeS:v}.
2016-09-27: Add entries for (uncertain) canon from Smithsonian GO FLIGHT app.
2016-08-30: Add canon from Smithsonian Air & Space Magazine interview.
2016-08-01: New words from qep'a' 2016.
2016-07-01: More minor edits to notes.
2016-04-13: Minor edits to notes.
2016-02-16: Fixed Planck time bug. Added many German definitions.
2015-12-29: add links from TKD and KGT references to Google Play Books. add KLI activities to menu.
2015-12-12: misc. edits
2015-11-18: make compatible with Chrome
2015-10-28: added new vocabulary from Saarbrücken qepHom 2015.
2015-09-24: added transcription of Marc Okrand's name.
2015-08-03: vocab for bodily gases.
2015-05-01: use Proguard for TTS version too.
2015-04-22: minor edits to database.
2015-04-07: minor edits to database.
2015-03-27: Update Play Store URL to {quja'} keyboard in help.
2015-03-23: Added Julius Caesar and full moon.
2015-03-06: Add new words from Festival of the Spoken Nerd DVD. Minor edits.
2015-02-06: Add Italian translations for UI.
2015-01-12: Remove tutorial (since ShowcaseView has compatibility issues with newer Android versions)
2014-12-25: Attempt to update appcompat support library to v21, failure.
2014-12-17: Remove extcan from DeCandido words vetted by Marc Okrand for qepHom 2014.
2014-12-03: Remove hyp annotation from some Klingon Monopoly words.
2014-12-02: Added colour-coding to indicate different parts of speech. Show {pIqaD} in more places.
2014-11-30: Fixed rover placement bug with some complex sentences.
2014-11-26: Added extra-canonical vocabulary from Klingon Blockly.
2014-11-25: Added social network preference setting.
2014-11-16: qepHom vocabulary related to: ranks, computers, knots, quantum mechanics, The Klingon Art of War.
2014-11-12: qepHom vocabulary related to geometry and card games.
2014-10-23: Audio complete for TTS engine.
2014-09-25: Misc. updates.
2014-07-27: Added new proverbs from qep'a' 2014. Enabled search for German definitions.
2014-07-26: Remove dependency on ActionBarSherlock.

AppCompat.Dark theme generated using:
http://jgilfelt.github.io/android-actionbarstylegenerator/#name=Appcompatdark&compat=appcompat&theme=dark&actionbarstyle=transparent&texture=0&hairline=0&neutralPressed=1&backColor=2D2D2D%2C100&secondaryColor=151515%2C100&tabColor=33B5E5%2C100&tertiaryColor=303030%2C100&accentColor=33B5E5%2C100&cabBackColor=002E3E%2C100&cabHighlightColor=33B5E5%2C100

2014-07-21: Link hand and foot words; improve Shakespeare entries; added lyrics for {yIH bom}.
2014-07-15: Try to work around a bug where the action bar sometimes disappears.
2014-06-14: Minor edits to database.
2014-04-29: Better integration with TTS.
2014-04-16: Add Text-To_Speech ability (requires TTS engine which supports Klingon).


KlingonAssistant/data/generate_db.sh
#!/bin/bash

# Check for non-interactive mode flag.
if [[ "$1" = "--noninteractive" ]]
then
    NONINTERACTIVE=true
    shift
fi

# Check for xml-only mode flag (exclusive with "--noninteractive").
if [[ "$1" = "--xmlonly" ]]
then
    XMLONLY=true
    shift
fi

# Concatenate data into one xml file.
cat mem-00-header.xml mem-01-b.xml mem-02-ch.xml mem-03-D.xml mem-04-gh.xml mem-05-H.xml mem-06-j.xml mem-07-l.xml mem-08-m.xml mem-09-n.xml mem-10-ng.xml mem-11-p.xml mem-12-q.xml mem-13-Q.xml mem-14-r.xml mem-15-S.xml mem-16-t.xml mem-17-tlh.xml mem-18-v.xml mem-19-w.xml mem-20-y.xml mem-21-a.xml mem-22-e.xml mem-23-I.xml mem-24-o.xml mem-25-u.xml mem-26-suffixes.xml mem-27-extra.xml mem-28-footer.xml > mem.xml

if [[ $XMLONLY ]]
then
    exit
fi

# Write database version number.
VERSION=$(cat VERSION)
echo Writing database version $VERSION...
sed -i "s/\[\[VERSION\]\]/$VERSION/" mem.xml

# Convert from xml to sql instructions.
./xml2sql.pl > mem.sql
sed -i 's/INSERT INTO "mem"/INSERT INTO mem/g' mem.sql

# Print any entries with duplicate columns.
grep "ARRAY" mem.sql

# Print any parts of speech accidentally entered into the definition.
POS_DEFINITION_MIXUP=$(grep -B2 "definition\">\(v\|n\|adv\|conj\|ques\|sen\|excl\)[:<]" mem.xml)
if [[ ! -z "$POS_DEFINITION_MIXUP" ]]
then
    echo "Part of speech information entered into definition:"
    echo "$POS_DEFINITION_MIXUP"
    echo
fi

# Print any empty German definitions.
MISSING_DE=$(grep -B3 "definition_de\"><" mem.xml)
if [[ ! -z "$MISSING_DE" ]]
then
    echo "Missing German definitions:"
    echo "$MISSING_DE"
    echo
fi

# Pause (in case of error).
if [[ ! $NONINTERACTIVE && (! -z "$POS_DEFINITION_MIXUP" || ! -z "$MISSING_DE") ]]
then
    read -n1 -r -p "Press any key to continue..."
    echo
fi

# Create db binary.
git checkout ../app/src/main/assets/qawHaq.db
if [[ -f ../app/src/main/assets/qawHaq.db ]]
then
    if [[ ! $NONINTERACTIVE ]]
    then
        # If the db already exists, show a diff.
        sqlite3 ../app/src/main/assets/qawHaq.db .dump > old-mem.sql
        sed -i 's/INSERT INTO "mem"/INSERT INTO mem/g' old-mem.sql
        vimdiff old-mem.sql mem.sql
        read -n1 -r -p "Press any key to generate new db..."
        echo
    fi
    mv ../app/src/main/assets/qawHaq.db ../app/src/main/assets/qawHaq.db~
fi
sqlite3 ../app/src/main/assets/qawHaq.db < mem.sql

# Sanity check.
sqlite3 ../app/src/main/assets/qawHaq.db .dump > sanity.sql
sed -i 's/INSERT INTO "mem"/INSERT INTO mem/g' sanity.sql
IN_OUT_DIFF=$(diff mem.sql sanity.sql)
if [[ ! -z "$IN_OUT_DIFF" ]]
then
    echo "Sanity check failed, entries possibly missing or out of order:"
    echo "$IN_OUT_DIFF"
    echo
fi

# Pause (in case of error).
if [[ ! $NONINTERACTIVE ]]
then
    read -n1 -r -p "Press any key to delete temporary files..."
    echo
fi

# Clean up temporary files.
rm mem.xml
rm mem.sql
rm mem_processed.xml
rm sanity.sql
rm -f old-mem.sql
rm -f ../app/src/main/assets/qawHaq.db~
